build.ps1

```diff
@@ -1,33 +1,177 @@
 ## Copyright (c) Microsoft Corporation.
 ## Licensed under the MIT License.
-
-[CmdletBinding()]
+# default behavior is build only
+[CmdletBinding(SupportsShouldProcess=$true)]
 param (
     [Parameter()]
     [string]
-    $Configuration = "Debug",
+    $Configuration = "Release",
+
+    [Parameter()]
+    [switch]$Publish,
+
+    [Parameter()]
+    [switch]$Test,
+
+    [Parameter()]
+    [switch]$NoBuild,
 
     [Parameter()]
-    [switch]
-    $Clean
+    [switch]$Clean
 )
 
-try {
-    Push-Location "$PSScriptRoot/src/code"
+$moduleName = "Microsoft.PowerShell.TextUtility"
+$projectRoot = $PSScriptRoot
+
+$srcRoot = Join-Path $ProjectRoot src
+$codeRoot = Join-Path $srcRoot code
+$tstRoot = Join-Path $ProjectRoot test
+$modManifest = Join-Path $srcRoot "${ModuleName}.psd1"
+
+
+function AssertFile {
+    param (
+        [Parameter(Mandatory=$true)]
+        [string]$Path
+    )
+
+    if (-not (Test-Path $Path)) {
+        throw "File not found: $Path"
+    }
+}
+
+function Assert {
+    param ([Parameter(Mandatory=$true)][bool]$Condition, [Parameter(Mandatory=$true)][string]$Message)
+    if (! $Condition) {
+        throw $Message
+    }
+}
+
+foreach ($path in $srcRoot, $codeRoot, $tstRoot, $modManifest) {
+    AssertFile $path
+}
+
+$modInfo = Import-PowerShellDataFile $modManifest
+$modVersion = $modInfo.ModuleVersion
+Assert ($null -ne $modVersion) "ModuleVersion is null"
+$pubRoot = "${PSScriptRoot}/out/${ModuleName}/${modVersion}"
+
+function Build-Assembly {
+    if (Test-ShouldBuild) {
+        try {
+            Push-Location $codeRoot
+            if($PSCmdlet.ShouldProcess($Configuration)) {
+                dotnet build --configuration $Configuration
+            }
+        }
+        finally {
+            Pop-Location
+        }
+    }
+    else {
+        Write-Verbose -Verbose "No changes to build"
+    }
+}
+
+function Test-ShouldBuild {
+    $latestSourceTime = Get-LatestSourceTime
+    $latestBuildTime = Get-LatestBuildTime
+    return ($latestBuildTime -lt $latestSourceTime)
```

Comment:
Isn't this mostly handled by msbuild already?

Reply:
sort of - I've been irked by this for some time. If you run `dotnet build` 2 times in succession, you still see all of the compilation output even though it does not actually compile.
I feel this is a better user experience.